### ISSUE TITLE
6-bit Qwen3.8-27B pack: embedded-MTP on/off output is not byte-identical under greedy decoding

### ISSUE THREAD

**Issue by terrazorc** [ddalcu/mlx-serve#206]
# [Issue draft] MLXServe 26.8.8 — 6-bit Qwen3.8-27B pack: embedded-MTP on/off output is not byte-identical under greedy decoding

*Draft for the MLXServe upstream tracker. Prepared 2026-08-16; measurements
from zeroM5 (Apple M5 Max, 128 GB, Mac17,6), MLXServe v26.8.8
(release tarball `mlx-serve-macos-arm64`, sha 15a08318…), model pack
`ddalcu/Qwen3.8-27B-MLX-Serve-6bit` (21 GB, pulled via `mlx-serve pull`,
in-checkpoint MTP head auto-loaded: "[mtp] loading in-checkpoint head",
draft lm_head requantized 3-bit/gs64 by the server).*

## Claim being tested

For a given model/checkpoint and server version, greedy decoding
(temperature 0, fixed seed) should be deterministic — i.e., toggling the
embedded-MTP/speculative path (which changes only *how* candidates are
verified, not *which* tokens are emitted) should yield **byte-identical**
output. Several MLXServe release notes / community reports state that
speculative serving is byte-identical to serial decoding.

**We cannot reproduce that guarantee on the 6-bit pack on Apple Silicon
(M5 Max).** MTP-on and MTP-off diverge at character 218.

## Reproduction

1. Pull the pack and serve it on a free port (we used the default 11234):

   ```
   cd ~/tools/mlxserve/mlx-serve-macos-arm64
   ./mlx-serve pull ddalcu/Qwen3.8-27B-MLX-Serve-6bit
   # serve with default settings (embedded MTP path available)
   ```

2. Run the same greedy completion twice against `http://127.0.0.1:11234/v1`,
   only difference being `enable_mtp` (MTP on vs off):

   ```
   python3 harness.py probe http://127.0.0.1:11234/v1 ddalcu/Qwen3.8-27B-MLX-Serve-6bit --think off --tag orchtest-off
   python3 harness.py probe http://127.0.0.1:11234/v1 ddalcu/Qwen3.8-27B-MLX-Serve-6bit --think on  --tag orchtest-on
   ```

   (Our `harness.py` in this run directory issues an OpenAI-compatible
   chat completion: `temperature=0`, `seed=424242`, `max_tokens=300`,
   thinking off/on via top-level `enable_thinking`.)

3. Diff the two response bodies.

## Expected

Byte-identical content between the MTP-on and MTP-off runs (greedy, same
seed, same weights — draft verification must not change the argmax path).

## Actual

Output **diverges at character 218**:

- MTP-on : `…quantization format…`
- MTP-off: `…quantization method…`

A single near-tie argmax flip, after which the streams diverge. Both sides
are coherent and correct content — this is a numerics/nondeterminism issue,
not corruption.

## Environment

| item | value |
|---|---|
| Node | Apple M5 Max, 128 GB (Mac17,6), macOS |
| Server | MLXServe v26.8.8 (`mlx-serve-macos-arm64`, tarball sha 15a08318…), embedded MLX 0.32.0 |
| Model pack | `ddalcu/Qwen3.8-27B-MLX-Serve-6bit` (21 GB, 6-bit, in-checkpoint MTP head) |
| Endpoint | `http://127.0.0.1:11234/v1` |
| Decoding | greedy: `temperature=0`, `seed=424242`, `max_tokens=300` |
| Toggle | `enable_mtp` on vs off (thinking on/off variant probed the same way) |
| Divergence | char 218 — "quantization format" (MTP on) vs "quantization method" (MTP off) |

## Context / cross-checks (helpful if upstream pins different conditions)

- The vendor release-test suite may pin conditions under which the
  byte-equality claim holds (different quant, different head quant, PLD
  off, specific ctx size). We record what we measured on this cell.
- Same near-lossless class observed independently on mlx-vlm 0.6.12
  (python) MTP lanes: 4-bit + MTP was byte-identical to serial on short/
  medium generations with rare near-tie flips on long ones; 8-bit + MTP
  flipped near char 33 on the same prompt. So "batched-verify numerics
  perturb near-tie argmax" appears to be a general property of speculative
  verification on these MLX builds, not unique to MLXServe — but MLXServe
  specifically markets the 6-bit pack with the byte-identical-greedy
  expectation, and our gate (exact-string probes) fails on it.
- PLD (prompt-local draft) is on by default in 26.8.8 and self-gates to
  neutral on novel content (verified: 26.7 vs 26.2 tok/s); if PLD
  participates in the MTP-on path, it may be a contributing nondeterminism
  source.

## Questions for maintainers

1. Is MTP-on/off byte-equality on the 6-bit pack a supported guarantee, or
   an informal property? If supported, under which conditions (PLD on/off,
   head quant, ctx size)?
2. Can the verify path be made deterministic (e.g., force-serial argmax
   compare, or a `deterministic_verify` flag) for gate-style consumers?
3. Which conditions does the release-test suite pin for the byte-identity
   claim?

## Environment addendum

- Repro commands reference a local harness; the essential call is any OpenAI-compatible chat completion at `temperature=0`, `seed=424242`, `max_tokens=300`, identical prompt, with `"enable_mtp": true` vs absent/false. Diff content.
- Server log at load: `[mtp] loading in-checkpoint head from the trunk shards` + `[mtp] draft-only lm_head requantized to 3-bit/gs64`.

**Comment by ddalcu** [comment]
First question: 
What are you practically hitting here? Your gate does exact-string probes, but both outputs are correct content, so I'd like to understand what the exact-match requirement buys you in practice. If the gate is checking model quality, byte equality is the wrong instrument for it.

To your questions:

1. It's not a guarantee and never was. The supported claim is that MTP verification uses the trunk head, so speculation changes how fast tokens come out, not which distribution they come from. Byte equality mostly holds, but a verify forward runs multiple rows through every matmul while serial runs one, MLX picks kernels by batch width, and different kernels sum in a different order. Near-tie argmaxes flip. Your char 218 "format" vs "method" is exactly that. Our own equivalence test acquits a divergence when the serial top-2 gap at that token is basically zero, and in live captures we've seen exact 0.0000 ties. This is the same property vLLM, TensorRT-LLM etc all have with batching, nobody fixes it because the fix is making every kernel's reduction order invariant to batch width, which forfeits exactly the shape-specialized kernels the spec win comes from.
2. The deterministic mode already exists: send enable_mtp: false per request, or launch with --no-mtp. Add --kv-quant off or 8 if you want byte-stable greedy across runs in general. A deterministic_verify flag would just be spec off with a different name, since even re-checking near-ties serially doesn't help: the KV written during a verify forward differs in the low bits from serially written KV, so the divergence lives in the cache, not just the logits.
3. The release suite pins the top-2-gap acquittal (a flip is only a failure if the serial gap was above ~0.15 nats), boots with the prefix cache off on hybrids, and asserts spec engagement from logs rather than output equality.

Side notes: seed does nothing at temperature 0, and PLD isn't involved, MTP has priority so PLD never runs when MTP is engaged. priority: MTP > dflash > drafter > PLD

I'll add the byte-stable recipe to the docs so this expectation is written down somewhere.
